Initial Project Configuration

Antypurus opened the pull request:
Closes #3

## Description 

This pull request configured the initial folder structure of the project, its build systems alongside scripts used for compilation, a base for testing integrated into the build system, and, configured the CircleCI CI/CD pipeline for the project.

## Build System

### Folder Structure
 - The `CI` folder contains scripts and utilities that are regularly used by the CI/CD pipeline.
 - The `scripts` folder contains scripts related to various aspects of the build system and utilities for the developer.
 - The `tests` folder contains the testing CMake configuration, the source code for the testing frameworks used, and the actual project tests source code.
     - The `src` subfolder contains the actual test source code.
     - The `ThirdParty` subfolder contains the source code used to build the test framework used. In this case catch2
 - The `src` folder contains the C++ source code for the project as well as individual modules cmakelists.
    - The subfolder `Engine` contains the source code for the engine
    - The subfolder `Common` contains code that is shared between the engine and the editor
    - The subfolder `Editor` contains the code for the editor that will be used alongside the engine. For the foreseeable future, this will be used as a test entry point for the engine code.
 - The `.circleci` contains the CircleCI configuration files 

### Cmake Scripts
This project makes heavy used of Cmake targets and subdirectories to structure itself. 

There is a global CMakelists at the root of the project, this script handles controlling if tests should be included in the build or not. Tests can be turned on and off by setting the value of the CMake option variable `ENABLE_TESTS`.

Inside of `src` is where most of the project subdivision happens, there is a single CMakelists at the root of the `src` directory which just includes the subdirectories for the engine and the editor. 

The engine is a cmake target that is compiled into a library, it can be compiled either into a static library or a dll and this can be done via the cmake variable `ENGINE_DYNAMIC_COMPILE`. In the case the engine is compiled as DLL some extra compiler macro defines are set so that the source code can know to toggle on dll exports and imports depending on the target. Source files in the engine directory are automatically detected and added to build during compilation, so no changes are required to the cmake script. Files in the Common folder are also automatically added to the build for the engine.

The editor is compiled into an actual executable. Files in the editor directory and common directory are automatically added to the editor build.

Both the engine and editor are configured to output their respective build files into a directory called `bin` in the build directory.

The test cmake configuration includes the editor, engine, and common code automatically so that it can be easily accessed to build tests. There is automatic test discovery so that CTest can easily be used. Binary files are output into a folder called `tests` inside of the `bin` folder in the build directory.

### Build Scripts
Build scripts are contained in the folder `scripts`. This folder is organized in compiler folders, one for `clang` and one for `msvc`. Each compiler folder has a `build` folder and a `tests` folder, used to build the project and the tests of the project respectively. Both the tests and build folder have the same layout, containing a Debug and Release folder used to build and run those configurations as well as a build and run file for the last known configuration for the system. All the scripts presented here output their result in `build/compiler` folder.
- `clean.bat` - deletes the entirety of the `build` folder.
- `config-clang-tidy.bat` - configures the project to run clang-tidy on it.
- `lint.bat` - runs source code style checking on the project via clang-format.
- `open-vs.bat` - open the project in visual studio via the generated `msvc` sollution file.
- `static-analysis.bat` - run static analysis on the project via clang-tidy.
- `compiler/clean.bat` - deletes the build folder associated with the compiler
- `compiler/cmake.bat` - configures the build folder and cmake for the specified compiler
- `compiler/build/run.bat` - runs the project in the last used configuration, recompiling if required.
- `compiler/build/build.bat` - compiles the project in the last used configuration.
- `compiler/build/_run_no_compile.bat` - runs the project in the last compiled configuration. Project recompilation will not happen with this script.
- `compiler/build/Configuraiton(Release or Debug)/build.bat` - builds the project in the specified configuration.
- `compiler/build/Configuraiton(Release or Debug)/build.bat` - runs the project in the specified configuration, recompiling if required.
- `compiler/tests/run.bat` - runs the tests in the last used configuration, recompiling if required.
- `compiler/tests/build.bat` - compiles the tests in the last used configuration.
- `compiler/tests/_run_no_compile.bat` - runs the tests in the last compiled configuration. Project recompilation will not happen with this script.
- `compiler/tests/Configuraiton(Release or Debug)/build.bat` - builds the tests in the specified configuration.
- `compiler/tests/Configuraiton(Release or Debug)/build.bat` - runs the tests in the specified configuration, recompiling if required.

## CI/CD Pipeline Structure
**Linting stage** -> **static analysis stage** -> **Compilation Test Stage** -> **Unit Test Stage** -> **Release Candidate Published on Github**

- `Linting Stage` - Linting is run via `clang-format`.
- `Static Analysis Stage` - Runs C++ static analysis via `clang-tidy`.
- `Compilation Stage`
    - `MSVC Debug Compilation` - Compiles the project in the Debug configuration via MSVC
    - `MSVC Release Compilation` - Compiles the project in the Release configuration via MSVC 
    - `Clang Debug Compilation` - Compiles the project in the Debug configuration via Clang
    - `Clang Release Compilation` - Compiles the project in the Release configuration via Clang
- `Unit Test Stage`
    - `MSVC Debug Unit Tets` - Runs the test in Debug mode compiled by MSVC
    - `MSVC Release Unit Tets` - Runs the test in Release mode compiled by MSVC
    - `Clang Debug Unit Tets` - Runs the test in Debug mode compiled by Clang 
    - `Clang Release Unit Tets` - Runs the test in Release mode compiled by Clang 
- `Release Candidate Published to Github` - Publishes a release candidate to the GitHub releases marked as pre-released. This stage of the CI is only run in commits to the  `dev` branch.

Antypurus committed:
started workin on a pipeline configuration

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
Add .circleci/config.yml

Antypurus committed:
WIP

Antypurus committed:
Finished Initial Version Of Continuous Integration Pipeline


WIP

WIP

WIP: added requirements file to run static analysis python script

WIP

WIP

WIP

WIP

WIP

WIP

WIP:ready to prep for merge

Antypurus committed:
WIP:started working on a unit testing pass

Antypurus committed:
WIP: working on test stage

Antypurus committed:
working on testing config

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
updateing project layout

Antypurus committed:
refactoring project structure

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
added script that open visual studio project

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
added clean scripts

Antypurus committed:
added super build file

Antypurus committed:
fixing clang output binary extension

Antypurus committed:
tool script documentation

Antypurus committed:
WIP

Antypurus committed:
cleaning up ci scripts

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
explicitly made file list strings

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
started fixing static analysis

Antypurus committed:
working on static analysis

Antypurus committed:
static analysis should now work well

Antypurus committed:
static analysis should now work well

Antypurus committed:
static analysis should now work well

Antypurus committed:
static analysis should now work well

Antypurus committed:
static analysis should now work well

Antypurus committed:
WIP

Antypurus committed:
WIP

Antypurus committed:
wip

Antypurus committed:
WIP

Antypurus committed:
Merge branch 'CI_Setup' of https://github.com/TiagoJoseMagalhaes/Liquium into CI_Setup

Antypurus commented on .circleci/config.yml:
```diff
@@ -0,0 +1,232 @@
+       - run:
+          name: "Install Cmake"
+          command: |
+                  choco install llvm ninja -y
```
msvc compilation does not need ninja build system

Antypurus commented on .circleci/config.yml:
```diff
@@ -0,0 +1,232 @@
+       - run:
+          name: "Install Cmake"
+          command: |
+                  choco install llvm ninja -y
```
msvc compilation boes not need ninja build system

Antypurus replied:
fixed

Antypurus commented on .circleci/config.yml:
```diff
@@ -0,0 +1,232 @@
+       - run:
+          name: "Install Cmake"
+          command: |
+                  choco install llvm ninja -y
```
msvc does not need llvm or ninja installed for running tests

Antypurus replied:
fixed

Antypurus commented on .circleci/config.yml:
```diff
@@ -0,0 +1,232 @@
+       - run:
+          name: "Install Cmake"
+          command: |
+                  choco install llvm ninja -y
```
msvc does not need llvm on ninja installed to run tests

Antypurus replied:
fixed

Antypurus commented on CI/Lint.bat:
```diff
@@ -0,0 +1,3 @@
+@echo off
```
needs directory independence code (loop that finds the .git folder which denotes the root of the project)

Antypurus replied:
fixed

Antypurus commented on CI/Linting/lint.py:
```diff
@@ -0,0 +1,38 @@
+    return clang_output.returncode
+
+def main():
+    files = fetch_all_file_to_lint("./src/Engine")
```
This need to check all of the source, not just the engine. Also, keep in mind that must also be done for the tests.

Antypurus replied:
fixed

Antypurus commented on CI/gen_version.py:
```diff
@@ -0,0 +1,28 @@
+rc_release_regex = re.compile('v([0-9]+\.[0-9]+\.[0-9]+)-rc([0-9]+)')
+
+#fetch latest git tag
+hash = subprocess.check_output(["git", "rev-list","--tags","--max-count=1"]).strip().decode("utf-8") 
+release_version = subprocess.check_output(["git","describe","--tags","--abbrev=0", hash]).strip().decode("utf-8") 
+
+#read patch version
+file = open('./VERSION','r')
+local_version = file.read();
```
These operations can fail so maybe adding some form of error checking would be good.

Antypurus replied:
wont fix

Antypurus commented on CMakeLists.txt:
```diff
@@ -1,26 +1,12 @@
-        "windowscodecs.lib"
-        )
\ No newline at end of file
+option(ENABLE_TESTS "Controls weather to compile tests or not" ON)
+# testing can only be done with msvc because google test is garbage
+if(ENABLE_TESTS)
+    enable_testing()
+    add_subdirectory(tests)
+endif()
```
As a discussion does we wanna have the tests setting always on once its set or more like something that has to be done explicitly when we want it (think what we do internally for cross-compilation and code upload)

Antypurus replied:
Honestly, I don't think it matters since it all managed through the scripts

Antypurus commented on scripts/clang/build/Debug/build.bat:
```diff
@@ -0,0 +1,48 @@
+cd build/clang
+ninja -j %NUMBER_OF_PROCESSORS%
+if errorlevel 1 (
+    echo Failed to compile code
```
maybe improve the logs in general for the scripts with color (I know its a pain but it would be good)

Antypurus replied:
not gonna do this anytime soon fuck scripts

Antypurus commented on scripts/clang/clean.bat:
```diff
@@ -0,0 +1,3 @@
+@echo off
```
lacks directory independence loop

Antypurus commented on scripts/config-clang-tidy.bat:
```diff
@@ -0,0 +1,5 @@
+call scripts/clang/cmake.bat -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DENABLE_TESTS=OFF
```
Disable clang-tidy inside of catch cmake files instead of globally disabling it for tests.

TiagoJoseMagalhaes replied:
This is significantly more troublesome than initially thought and not really worth much since it only affects tests.

Antypurus replied:
fixed

Antypurus commented on scripts/lint.bat:
```diff
@@ -0,0 +1,3 @@
+@echo off
+
+call ./CI/Lint.bat
```
return error code from this call

Antypurus commented on scripts/msvc/clean.bat:
```diff
@@ -0,0 +1,3 @@
+@echo off
```
lack directory independence loop

Antypurus commented on scripts/open-vs.bat:
```diff
@@ -0,0 +1,17 @@
+Rem this script must be called from the liquium project root
+Rem this script runs the cmake script and open the visual studio project
+
+cmake -Bbuild/msvc .
```
use the existing script for configuring msvc cmake build

Antypurus commented on scripts/open-vs.bat:
```diff
@@ -0,0 +1,17 @@
+@echo off
+Rem this script must be called from the liquium project root
+Rem this script runs the cmake script and open the visual studio project
+
```
lack directory independence code

Antypurus commented on scripts/run-static-analysis.bat:
```diff
@@ -0,0 +1,10 @@
+@echo off
+Rem this script should be called from the liquium project root
+Rem this script will run static analysis via clang tidy
+
```
lack directory independence code.

Antypurus commented on scripts/run-static-analysis.bat:
```diff
@@ -0,0 +1,10 @@
+Rem this script will run static analysis via clang tidy
+
+set "directory=%cd%"
+call scripts/config-clang-tidy.bat
```
handle error scenario from this script

Antypurus commented on scripts/run-static-analysis.bat:
```diff
@@ -0,0 +1,10 @@
+Rem this script should be called from the liquium project root
+Rem this script will run static analysis via clang tidy
+
+set "directory=%cd%"
```
i dont think this is needed so remove it

Antypurus commented on src/Editor/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+set(LIQUIUM_EDITOR "Editor")
```
could add file glob for automatic source file discovery but I guess it's not need for now.

Antypurus commented on src/Engine/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+        set(warnings "/W4 /WX /EHsc")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /wd26812")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /MP /nologo /Oi")
```
intrinsic optimization should only be on in release builds not in all builds.

Antypurus commented on src/Engine/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+        *.cpp
+        *.hpp)
+
+add_library(Liquium STATIC ${ENGINE_SOURCES})
+
+set_target_properties(Liquium PROPERTIES PREFIX "")
+set_target_properties(Liquium PROPERTIES SUFFIX ".lib")
```
need some way to switch between static and dynamic compilation/linking

Antypurus reviewed (commented):
Here is what needs to get fixed

Antypurus committed:
WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

IWP

IWP

IWP

IWP

IWP

IWP

IWP

IWP

IWP

IWP

IWP

IWP

WIP

wip

WIP

WIP

WIP

WIP

WIP

test

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

wip

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

testing new cmake config

WIP

starting to get somewhere

WIP

WIP

WIP

working on testing

started working on testing system improvements

WIP

stuff

WIP

i dont think clang will work, too bad

started refactoring toolscripts

refactoring toolscripts

WIP

wip

wip

wip

WIP

WIP

WIP

started working on the toolscript refactor

updating scripts

removed unnecessary variable

removed unnecessary variable

added clang release build script

added rebuld and buld scripts for msvc

added rebuld and buld scripts for msvc

started working on tests again

started working on test scripts

started working on testing scripts

added raw catch2 so its easier to integrate with cmake

reconfigured the tests

test

seems i was able to get it working

no longer need to use the msbuild llvm toolchian thingy that i dont like

i dont think i need all this crap tbh

working on script to run tests in debug mode

turns out the debug scripts were not being tracked

working on tests scripts

working on tests scripts

working on tests

added files for testing under msvc

working on build scripts

fixed to build type detenction in run scripts

some more fixes

some more fixes

some more fixes

some more fixes

some more fixes

finally done refactoring build scripts

new ci config

lint fixes

lint fixes

fixed static analysis

fixed static analysis

fixed some ci issues

fixed some ci issues

fixed some ci issues

fixed some ci issues

fixed tests

started testing github releasing

started testing github releasing

started testing github releasing

started testing github releasing

files for versioning

files for versioning

files for versioning

files for versioning

files for versioning

files for versioning

updated version generator to handle local changes to version

added branch shit

Revert "added branch shit"

This reverts commit 61562d6d712370bb4b3c94d6621b5c538f31b1fa.

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

Updated config.yml
trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

trying to get branch filtering for circle ci working

some fixes to the ci config

some fixes

fixes

fixes

fixed tests

added dll compile switch

added dll compile switch

added dll compile switch

added dll compile switch

fix related to dll modification to build system

some fixes to clang tidy

fixes

fixes

test

test

test

test

Antypurus committed:
last minute fixes

Antypurus committed:
set version ahead of merge to dev
